doc/Configuration.md
# Configuration

The following table lists the configurable parameters of the `Vault Sidecar Injector` chart and their default values.

| Parameter    | Description          | Default                                                         |
|:-------------|:---------------------|:----------------------------------------------------------------|
| image.applicationNameLabel   | Application Name. Must match label com.talend.application | talend-vault-sidecar-injector   |
| image.metricsPort                | Port exposed for metrics collection | 9000 |
| image.path       | Image path   | talend/vault-sidecar-injector |
| image.port       | Service main port    | 8443            |
| image.pullPolicy   | Pull policy for image: IfNotPresent or Always       | IfNotPresent           |
| image.serviceNameLabel   | Service Name. Must match label com.talend.service     | talend-vault-sidecar-injector      |
| image.tag  | Image tag     | latest *(local testing)*, [VERSION_VSI](../VERSION_VSI) *(release)* |
| imageRegistry  | Image registry |   |
| injectconfig.jobbabysitter.image.path   | Image path | everpeace/curl-jq |
| injectconfig.jobbabysitter.image.pullPolicy | Pull policy for image: IfNotPresent or Always | Always |
| injectconfig.jobbabysitter.image.tag   | Image tag  | latest |
| injectconfig.jobbabysitter.resources.limits.cpu | Job babysitter sidecar CPU resource limits | 120m |
| injectconfig.jobbabysitter.resources.limits.memory | Job babysitter sidecar memory resource limits | 25Mi |
| injectconfig.jobbabysitter.resources.requests.cpu | Job babysitter sidecar CPU resource requests | 100m |
| injectconfig.jobbabysitter.resources.requests.memory | Job babysitter sidecar memory resource requests | 20Mi |
| injectconfig.vault.image.path  | Image path  | vault |
| injectconfig.vault.image.pullPolicy    | Pull policy for image: IfNotPresent or Always  | Always   |
| injectconfig.vault.image.tag  | Image tag | 1.6.5 |
| injectconfig.vault.log.format                    | Vault log format: standard, json    | json    |
| injectconfig.vault.log.level                    | Vault log level: trace, debug, info, warn, err    | info    |
| injectconfig.vault.resources.limits.cpu | Vault sidecar CPU resource limits | 200m |
| injectconfig.vault.resources.limits.memory | Vault sidecar memory resource limits | 80Mi |
| injectconfig.vault.resources.requests.cpu | Vault sidecar CPU resource requests | 160m |
| injectconfig.vault.resources.requests.memory | Vault sidecar memory resource requests | 65Mi |
| mutatingwebhook.annotations.appLabelKey | Annotation for application's name. Annotation's value used as Vault role by default. | com.talend.application  |
| mutatingwebhook.annotations.appServiceLabelKey | Annotation for service's name | com.talend.service  |
| mutatingwebhook.annotations.keyPrefix | Prefix used for all vault sidecar injector annotations | sidecar.vault.talend.org  |
| mutatingwebhook.cert.cacertfile | Default filename for webhook CA certificate (PEM-encoded) in generated or provided Kubernetes Secret | ca.crt |
| mutatingwebhook.cert.certfile | Default filename for webhook certificate (PEM-encoded) in generated or provided Kubernetes Secret | tls.crt |
| mutatingwebhook.cert.certlifetime | Default lifetime in years for generated certificates. Not used if generated is false. | 10 |
| mutatingwebhook.cert.generated | Controls whether webhook certificates, private key and Kubernetes Secret are generated. If not, you have to provide a Kubernetes Secret with name secretName. | true |
| mutatingwebhook.cert.keyfile | Default filename for webhook private key (PEM-encoded) in generated or provided Kubernetes Secret | tls.key |
| mutatingwebhook.cert.secretName | Name of the Kubernetes Secret that contains the webhook certificates and private key. Secret should be in webhook's namespace. To provide if generated is false. | talend-vault-sidecar-injector-cert |
| mutatingwebhook.failurePolicy | Defines how unrecognized errors and timeout errors from the admission webhook are handled. Allowed values are Ignore or Fail | Ignore |
| mutatingwebhook.loglevel | Enable V-leveled logging at the specified level | 4 |
| mutatingwebhook.namespaceSelector.boolean    | Enable to control, with label "vault-injection=enabled", the namespaces where injection is allowed (if false: all namespaces except _kube-system_ and _kube-public_) | false                                                           |
| mutatingwebhook.namespaceSelector.namespaced | Enable to control, with label "vault-injection={{ .Release.Namespace }}", the specific namespace where injection is allowed (ie, restrict to namespace where injector is installed) | false |
| probes.liveness.failureThreshold                | Number of probe failure before restarting the probe                                 | 3  |
| probes.liveness.initialDelaySeconds             | Number of seconds after the container has started before the probe is initiated     | 2  |
| probes.liveness.periodSeconds                   | How often (in seconds) to perform the probe                                         | 20 |
| probes.liveness.timeoutSeconds                  | Number of seconds after which the probe times out                                   | 5  |
| probes.readiness.failureThreshold               | Number of probe failure before setting the probe to Unready                         | 3  |
| probes.readiness.initialDelaySeconds            | Number of seconds after the container has started before the probe is initiated     | 2  |
| probes.readiness.periodSeconds                  | How often (in seconds) to perform the probe       | 20   |
| probes.readiness.successThreshold      | Minimum consecutive successes for the probe to be considered successful after having failed  | 1  |
| probes.readiness.timeoutSecon          | Number of seconds after which the probe times out  | 5   |
| registryKey         | Name of Kubernetes secret for image registry                        |  |
| replicaCount                        | Number of replicas | 3    |
| resources.limits.cpu                | CPU resource limits                             | 250m         |
| resources.limits.memory             | Memory resource limits                          | 256Mi        |
| resources.requests.cpu              | CPU resource requests                           | 100m         |
| resources.requests.memory           | Memory resource requests                        | 128Mi        |
| revisionHistoryLimit                | Revision history limit in tiller / helm / k8s   | 3            |
| service.exposedServicePort   | Port exposed by the K8s service (Kubernetes always assumes port 443 for webhooks) | 443 |
| service.name                                    | Service name            | talend-vault-sidecar-injector                                   |
| service.prefixWithHelmRelease                   | Service name to be prefixed with Helm release name                                       | false                                                           |
| service.type                        | Kubernetes service type: ClusterIP, NodePort, LoadBalancer, ExternalName  | ClusterIP    |
| vault.addr                          | Address of Vault server    | `null` - To be provided at deployment time (e.g.: https://vault:8200)   |
| vault.authMethods.approle.path      | Path defined for AppRole Auth Method            | approle |
| vault.authMethods.approle.roleid_filename    | Filename for role id    | approle_roleid   |
| vault.authMethods.approle.secretid_filename  | Filename for secret id  | approle_secretid |
| vault.authMethods.kubernetes.path      | Path defined for Kubernetes Auth Method            | kubernetes |
| vault.ssl.verify               | Enable or disable verification of certificates               | true |

You can override these values at runtime using the `--set key=value[,key=value]` argument to `helm install`. For example,

```bash
$ helm install vault-sidecar-injector \
               <chart_folder_location> \
               --namespace <your namespace> \
               --set <parameter1>=<value1>,<parameter2>=<value2>
```


CHANGELOG.md
# Changelog for Vault Sidecar Injector

## Release v7.2.1 - 2021-05-26

Minor release to provide smaller Vault Sidecar Injector image and new default Vault image to fix CVE-2021-32923 (refer to HashiCorp Vault's [changelog](https://github.com/hashicorp/vault/blob/master/CHANGELOG.md#may-20th-2021-1)).

**Changed**

- [VSI #53](https://github.com/Talend/vault-sidecar-injector/pull/53) - Rework Dockerfile to reduce image size
- [VSI #52](https://github.com/Talend/vault-sidecar-injector/pull/52) - Update HashiCorp Vault image to 1.6.5

## Release v7.2.0 - 2021-05-19

This release comes with support for `admission.k8s.io/v1` AdmissionReview and `admissionregistration.k8s.io/v1` MutatingWebhookConfiguration on Kubernetes 1.16+. As a result, Vault Sidecar Injector now handles both v1 and v1beta1 versions of those resources.

*Note that `admission.k8s.io/v1beta1` AdmissionReview and `admissionregistration.k8s.io/v1beta1` MutatingWebhookConfiguration should not be supported (nor available) anymore on Kubernetes 1.22+*

**Changed**

- [VSI #48](https://github.com/Talend/vault-sidecar-injector/pull/48) - Minor chart updates (adjust CPU & memory for injected containers, add checks during chart install)
- [VSI #51](https://github.com/Talend/vault-sidecar-injector/pull/51) - Update base image to CentOS 7.9.2009

**Added**

- [VSI #49](https://github.com/Talend/vault-sidecar-injector/pull/49) - Add support for `admission.k8s.io/v1` AdmissionReview and `admissionregistration.k8s.io/v1` MutatingWebhookConfiguration (in addition to v1beta1)

## Release v7.1.1 - 2021-04-02

**Fixed**

- [VSI #45](https://github.com/Talend/vault-sidecar-injector/pull/45) - Do not inject sidecars if only job and secrets modes are enabled and secrets type is "static"

## Release v7.1.0 - 2021-02-09

A new `sidecar.vault.talend.org/vault-image` annotation has been added to override the default injected image. Refer to the [samples](samples) for a [working example](samples/app-dep-10-secrets_custom_image_notify.yaml).

The default Vault image has been bumped to version `1.6.2`.

**Changed**

- [VSI #44](https://github.com/Talend/vault-sidecar-injector/pull/44) - Update HashiCorp Vault image to 1.6.2

**Added**

- [VSI #43](https://github.com/Talend/vault-sidecar-injector/pull/43) - New annotation to allow injection of custom Vault image

## Release v7.0.2 - 2020-11-09

**Changed**

- [VSI #40](https://github.com/Talend/vault-sidecar-injector/pull/40) - Update doc references to Helm Hub (now Artifact Hub)

**Added**

- [VSI #41](https://github.com/Talend/vault-sidecar-injector/pull/41) - New configuration value to set Vault Agent log format

## Release v7.0.1 - 2020-10-28

Minor release with new default Vault image to address several security issues in HashiCorp Vault (CVE-2020-25816, CVE-2020-16250, CVE-2020-16251, CVE-2020-17455).

Note that provided chart now requires Helm v3.

**Changed**

- [VSI #34](https://github.com/Talend/vault-sidecar-injector/pull/34) - Documentation improvements
- [VSI #36](https://github.com/Talend/vault-sidecar-injector/pull/36) - Update HashiCorp Vault image to 1.5.4
- [VSI #37](https://github.com/Talend/vault-sidecar-injector/pull/37) - Update base image to CentOS 7.8.2003
- [VSI #38](https://github.com/Talend/vault-sidecar-injector/pull/38) - Migrate chart to Helm v3
- [VSI #39](https://github.com/Talend/vault-sidecar-injector/pull/39) - Code refactoring

## Release v7.0.0 - 2020-06-21

New with this release: Vault secrets can now be injected right into environment variables. Only **static** secrets are supported at the moment. Along with this feature, the `secrets` volumeMount is now automatically injected in your containers if not already defined (as it was the case with the `secrets` volume).

Default Vault image is now set to `1.4.2` to fix several CVEs (CVE-2020-13223, CVE-2020-12757: see HashiCorp's [CHANGELOG](https://github.com/hashicorp/vault/blob/master/CHANGELOG.md#142-may-21st-2020)).

Finally, the webhook certificates are no more generated using the Kubernetes Certificates API. The certificates and associated private key are now either generated by code or read from a Kubernetes Secret you provide. Refer to the updated [documentation](https://github.com/Talend/vault-sidecar-injector/blob/master/doc/Deploy.md#webhook-certificates) for details.

**Changed**

- [VSI #29](https://github.com/Talend/vault-sidecar-injector/pull/29) - Update HashiCorp Vault image to 1.4.2
- [VSI #30](https://github.com/Talend/vault-sidecar-injector/pull/30) - Webhook certificates generated by code or provided
- [VSI #31](https://github.com/Talend/vault-sidecar-injector/pull/31) - Migrate to Golang 1.14

**Added**

- [VSI #32](https://github.com/Talend/vault-sidecar-injector/pull/32) - Inject secrets in env vars. Feature announcement [here](https://github.com/Talend/vault-sidecar-injector/blob/master/doc/announcements/Injecting-secrets-in-env.md).

## Release v6.1.0 - 2020-05-18

This release fixes VSI deployment on Kubernetes 1.18+ clusters. It also comes with better AppRole integration and updated Vault image.

**Changed**

- [VSI #27](https://github.com/Talend/vault-sidecar-injector/pull/27) - Update HashiCorp Vault image to 1.4.1

**Added**

- [VSI #26](https://github.com/Talend/vault-sidecar-injector/pull/26) - Improve AppRole support: add tests, enforce check over secrets type, tune Vault Agent config

**Fixed**

- [VSI #25](https://github.com/Talend/vault-sidecar-injector/pull/25) - Fix RBAC following breaking change in [Kubernetes 1.18 Certificates API](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md). See also associated PR [86476](https://github.com/kubernetes/kubernetes/pull/86476) & [86933](https://github.com/kubernetes/kubernetes/pull/86933).

## Release v6.0.1 - 2020-04-06


cmd/vaultinjector-webhook/flags.go
// Copyright © 2019-2020 Talend - www.talend.com
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"flag"
	"fmt"
	"os"

	"k8s.io/klog"
)

func parseFlags() string {
	// Cert command parameters
	certCmd := flag.NewFlagSet(CertCmd, flag.ExitOnError)
	certCmd.StringVar(&certParameters.CertOperation, "certop", CreateCert, "operation on webhook certificates (create, delete)")
	certCmd.StringVar(&certParameters.CertSecretName, "certsecretname", "", "name of generated or provided Kubernetes secret storing webhook certificates and private key")
	certCmd.StringVar(&certParameters.CertHostnames, "certhostnames", "", "host names to register in webhook certificate (comma-separated list)")
	certCmd.IntVar(&certParameters.CertLifetime, "certlifetime", 10, "lifetime in years for generated certificates")
	certCmd.StringVar(&certParameters.CACertFile, "cacertfile", "ca.crt", "default filename for webhook CA certificate (PEM-encoded) in generated or provided k8s secret")
	certCmd.StringVar(&certParameters.CertFile, "certfile", "tls.crt", "default filename for webhook certificate (PEM-encoded) in generated or provided k8s secret")
	certCmd.StringVar(&certParameters.KeyFile, "keyfile", "tls.key", "default filename for webhook private key (PEM-encoded) in generated or provided k8s secret")

	// Webhook command parameters
	webhookCmd := flag.NewFlagSet(WebhookCmd, flag.ExitOnError)
	webhookCmd.IntVar(&webhookParameters.Port, "port", 8443, "webhook server port")
	webhookCmd.IntVar(&webhookParameters.MetricsPort, "metricsport", 9000, "metrics server port (Prometheus)")
	webhookCmd.StringVar(&webhookParameters.CACertFile, "cacertfile", "ca.crt", "PEM-encoded webhook CA certificate")
	webhookCmd.StringVar(&webhookParameters.CertFile, "certfile", "tls.crt", "PEM-encoded webhook certificate used for TLS")
	webhookCmd.StringVar(&webhookParameters.KeyFile, "keyfile", "tls.key", "PEM-encoded webhook private key used for TLS")
	webhookCmd.StringVar(&webhookParameters.WebhookCfgName, "webhookcfgname", "", "name of MutatingWebhookConfiguration resource")
	webhookCmd.StringVar(&webhookParameters.AnnotationKeyPrefix, "annotationkeyprefix", "sidecar.vault", "annotations key prefix")
	webhookCmd.StringVar(&webhookParameters.AppLabelKey, "applabelkey", "application.name", "key for application label")
	webhookCmd.StringVar(&webhookParameters.AppServiceLabelKey, "appservicelabelkey", "service.name", "key for application's service label")
	webhookCmd.StringVar(&webhookParameters.InjectionCfgFile, "injectioncfgfile", "", "file containing the mutation configuration (initcontainers, sidecars, volumes, ...)")
	webhookCmd.StringVar(&webhookParameters.ProxyCfgFile, "proxycfgfile", "", "file containing Vault proxy configuration")
	webhookCmd.StringVar(&webhookParameters.TemplateBlockFile, "tmplblockfile", "", "file containing the template block")
	webhookCmd.StringVar(&webhookParameters.TemplateDefaultFile, "tmpldefaultfile", "", "file containing the default template")
	webhookCmd.StringVar(&webhookParameters.PodLifecycleHooksFile, "podlchooksfile", "", "file containing the lifecycle hooks to inject in the requesting pod")

	if len(os.Args) == 1 {
		usage(os.Args[0])
		os.Exit(1)
	}

	vsiSubcommand := os.Args[1]

	switch vsiSubcommand {
	case CertCmd:
		klog.InitFlags(certCmd)
		certCmd.Parse(os.Args[2:])
	case WebhookCmd:
		klog.InitFlags(webhookCmd)
		webhookCmd.Parse(os.Args[2:])
	case VersionCmd:
		fmt.Println("VSI (Vault Sidecar Injector) version " + VERSION)
		os.Exit(0)
	default:
		usage(os.Args[0])
		os.Exit(1)
	}

	return vsiSubcommand
}

func usage(program string) {
	fmt.Printf("Usage: %s <command> [<args>]\n\nCommands:\n", program)
	fmt.Printf("  %s\n", CertCmd)
	fmt.Printf("  %s\n", WebhookCmd)
	fmt.Printf("  %s\n", VersionCmd)
	fmt.Printf("\nUse \"%s <command> --help\" for more information about a given command.\n", program)
}
